pkg/scheduler/plugins/gpushare/pod_handler.go

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2022 The Koordinator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gpushare
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	quotav1 "k8s.io/apiserver/pkg/quota/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/api/v1/resource"
+
+	apiext "github.com/koordinator-sh/koordinator/apis/extension"
+	schedulingv1alpha1 "github.com/koordinator-sh/koordinator/apis/scheduling/v1alpha1"
+)
+
+func (n *nodeDeviceCache) onPodAdd(obj interface{}) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		return
+	}
+	podRequest, _ := resource.PodRequestsAndLimits(pod)
+	if isGPUPod(podRequest) {
+		n.assume(nil, pod, pod.Spec.NodeName)
+	}
+}
+
+func (n *nodeDeviceCache) onPodUpdate(oldObj, newObj interface{}) {
+	return
+}
+
+func (n *nodeDeviceCache) onPodDelete(obj interface{}) {
+	var pod *corev1.Pod
+	switch t := obj.(type) {
+	case *corev1.Pod:
+		pod = t
+	case cache.DeletedFinalStateUnknown:
+		var ok bool
+		pod, ok = t.Obj.(*corev1.Pod)
+		if !ok {
+			return
+		}
+	default:
+		return
+	}
+	podRequest, _ := resource.PodRequestsAndLimits(pod)
+	if isGPUPod(podRequest) {
+		n.unAssume(nil, pod, pod.Spec.NodeName)
+	}
+}
+
+// add is used to update deviceUsed when there is a new pod created
+func (n *nodeDeviceCache) assume(allocations []*apiext.DeviceAllocation, pod *corev1.Pod, nodeName string) {
+	n.nodeDeviceInfos[nodeName].lock.Lock()
```

Comment:
how can you sure nodeName must exist in n.nodeDeviceInfos? you should use getNodeDevice to judge whether node exist. also in failover mode, you watch pod first, so you should consider create nodeDevice anyway.